### PR TITLE
Restrict code upload

### LINKED ISSUE #164: Control who can upload code
## Summary

As part of the governance meta-issue #161, we wish to have the ability to control who can upload wasm Code to the chain. Besides a governance proposal, this may be `Nobody | Somebody(Address) | Everybody`. The current default setting is `Everybody`.

## Problem Definition

If we wish to have a permissioned chain and ensure only high-quality, audited contracts suited for our domain are uploaded, the most obvious gatekeeper is to control the uploading of code in the first place. We set some module level `Params` that can configure this setting.

## Proposal

* Define some `Params` for `x/wasm` that contain an enum for who can upload code: `Nobody | Somebody(Address) | Everybody` (see #163 for enum def). Also a second enum for the default "instantiate code" permission if not specified on a message.
* Store/Load params in genesis
* Default value is `Everybody` for both fields (if not specified in genesis), to make this transparent for permissionless chains.
* Enforce access controls on `MsgStoreCode` (but not `StoreCodeProposal`, which is always legal)
* Test that we can modify this with a governance proposal ("Params change proposal") and add any helpers if needed to ensure we can.

```go
type WasmParams struct {
    UploadCodePerm Permission
    InstantiateContractPerm Permission
}
```
____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by alpe**
Resolves #164
______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

**Comment by alpe** [comment]
Sorry for the `push -f` hick up. My branch was not rebased proper.

File: x/wasm/internal/keeper/keeper.go [hunk 1]
```diff
@@ -93,7 +129,11 @@ func (k Keeper) Create(ctx sdk.Context, creator sdk.AccAddress, wasmCode []byte,
 	}
 	store := ctx.KVStore(k.storeKey)
 	codeID = k.autoIncrementID(ctx, types.KeyLastCodeID)
-	codeInfo := types.NewCodeInfo(codeHash, creator, source, builder)
+	if instantiateAccess == nil {
+		defaultAccessConfig := k.getInstantiateAccessConfig(ctx).With(creator)
+		instantiateAccess = &defaultAccessConfig
+	}
+	codeInfo := types.NewCodeInfo(codeHash, creator, source, builder, *instantiateAccess)
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/keeper.go:136]: This InstantiateInfo doesn't seem to be read anywhere.

I guess that change (permission on instantiate) is for another PR?

File: x/wasm/internal/keeper/genesis.go [hunk 2]
```diff
@@ -46,6 +46,7 @@ func InitGenesis(ctx sdk.Context, keeper Keeper, data types.GenesisState) error
 	if keeper.peekAutoIncrementID(ctx, types.KeyLastInstanceID) <= uint64(maxContractID) {
 		return sdkerrors.Wrapf(types.ErrInvalid, "seq %s must be greater %d ", string(types.KeyLastInstanceID), maxContractID)
 	}
+	keeper.setParams(ctx, data.Params)
```
[the remark this answers is not in the document]
Comment (alpe) [reply x/wasm/internal/keeper/genesis.go:49]: Params are normally set from the genesis. The validation is called by `module.	ValidateGenesis()` and implemented in `func (s GenesisState) ValidateBasic() error `

File: x/wasm/internal/keeper/keeper.go (hunk 1 above)
Comment (alpe) [reply x/wasm/internal/keeper/keeper.go:136]: yes, will push the other PR in a minute

File: x/wasm/internal/keeper/keeper_test.go [hunk 3]
```diff
@@ -44,7 +44,7 @@ func TestCreate(t *testing.T) {
 	wasmCode, err := ioutil.ReadFile("./testdata/contract.wasm")
 	require.NoError(t, err)
 
-	contractID, err := keeper.Create(ctx, creator, wasmCode, "https://github.com/CosmWasm/wasmd/blob/master/x/wasm/testdata/escrow.wasm", "cosmwasm-opt:0.5.2")
+	contractID, err := keeper.Create(ctx, creator, wasmCode, "https://github.com/CosmWasm/wasmd/blob/master/x/wasm/testdata/escrow.wasm", "cosmwasm-opt:0.5.2", nil)
```
[the remark this answers is not in the document]
Comment (alpe) [reply x/wasm/internal/keeper/keeper_test.go:47]: Should I replace the string with a more generic pattern? `foo/bar:latest`

File: x/wasm/internal/keeper/genesis.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/wasm/internal/keeper/genesis.go:49]: I keep forgetting Validation is somewhere else and assumed to be called before any functions.

File: x/wasm/internal/keeper/keeper_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/wasm/internal/keeper/keeper_test.go:47]: Sure, a valid but non-real string there is best and less likely to be outdated.

You can merge this PR without those changes, just a minor notice that should get fixed sometime.

[the remark this answers is not in the document]
Comment (alpe) [reply x/wasm/internal/keeper/keeper_test.go:47]: 👌 will keep this in mind next time I touch any related code.

**Pull request merged**